grin/src/Transformations/Names.hs

```diff
@@ -43,6 +43,19 @@ deriveNewName name = do
     then deriveNewName name
     else pure newName
 
+deriveWildCard :: NameM Name
+deriveWildCard = do
+  (newWildCard, conflict) <- state $ \env@NameEnv{..} ->
+    let wildcard = "_"
+        idx = Map.findWithDefault 0 wildcard namePool
+        new = packName $ printf "%s%d" wildcard idx
+    in  ( (new, Set.member new nameSet)
+        , env {namePool = Map.insert wildcard (succ idx) namePool, nameSet = Set.insert new nameSet}
+        )
+  if conflict
+    then deriveWildCard
```

Comment:
Why it can not be derive a wildcard with a new name in one go? Why the recursion is necessary?

Reply:
This code is a copy of the original name derivng algorithm. I assume we need the recursion because there could be some "already indexed" instances of the name we want to index. For example, we want derive a new wildcard for this program: 

```
grinMain = 
  _0 <- pure 0
  pure ()
```
First, we will derive the new name `_0`, increment the index for the name `_`, and then check whether `_0` is a member of the name set. It is, so we have to recurse. 

At least, that's how I understad it.